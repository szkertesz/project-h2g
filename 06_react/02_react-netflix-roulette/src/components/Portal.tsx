import { useState, useLayoutEffect } from 'react';
import { createPortal } from 'react-dom';

interface Props {
    children?: JSX.Element;
    containerId?: string;
}

export const createContainerAndAppendToBody = (containerId: string) =>{
    const containerElement = document.createElement('div');
    containerElement.setAttribute('id', containerId);
    document.body.appendChild(containerElement);
    return containerElement;
}

const Portal: React.FC<Props> = ({
    children,
    containerId = 'portal-container',
}) => {
    const [containerElement, setContainerElement] = useState<HTMLElement | null>(null);

    useLayoutEffect(() => {
        let systemCreated = false;
        let element = document.getElementById(containerId) as HTMLElement;
        if (!element) {
            systemCreated = true
            element = createContainerAndAppendToBody(containerId);
        }

        setContainerElement(element);

        return() => {
            if (systemCreated && element.parentNode) {
                element.parentNode.removeChild(element)
            }
        }
    }, [containerId]);

    if (containerElement === null) return null;

    return createPortal(children, containerElement);
};
export default Portal;
