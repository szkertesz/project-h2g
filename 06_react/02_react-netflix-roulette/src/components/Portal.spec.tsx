import { render, screen } from '@testing-library/react';
import Portal, {createContainerAndAppendToBody} from "./Portal";

describe('Portal', () => {
    it('should render component', () => {
        const containerId = 'test-portal-container';
        const { container } = render(<Portal containerId={containerId} />);
        screen.debug();
        expect(container).toMatchSnapshot();
    });
    it('should render component', () => {
        createContainerAndAppendToBody('test-portal-container')
        const { container } = render(<Portal containerId={'test-portal-container'}/>);
        screen.debug();
        expect(container).toMatchSnapshot();
    });
})
