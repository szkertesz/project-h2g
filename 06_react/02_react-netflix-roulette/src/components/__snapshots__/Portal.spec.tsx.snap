// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal should render component 1`] = `<div />`;

exports[`Portal should render component 2`] = `<div />`;
